en/Data Structures/Tries/trie.md

```diff
@@ -0,0 +1,161 @@
+However, if we represent the storage of numbers in a tree such that each letter appears only once in a particular level in the tree, we can achieve a much better search time
+
+```
+        e
+     /  |  \
+    a   n   g
+   / \  |   |
+  r   t d   g
+
+```
+
+You can see from the above tree representation, that all the words are in the tree, starting from the letter e, which starts all the words, then a, n, and g coming next and so on...
+The above representation is called a trie.
+
+# Standard Trie Operations 
+
+1) insert(): inserts the string into the trie.
+2) search(): searches for the string within the trie.
+
+# Building a Trie
+
+## Building a node for the elements of the trie
+
+To start building a trie, you first need to define a node with the revelant attributes needed for any trie.
+
+```
+class Node:
+    def __init__(self, val: string=None, is_word: bool=False):
+        self.val = val
+        self.is_word = is_word
+        self.children = {}
+```
+
+Here, you can see that the class `Node` has three instance attributes:
+1. val: *string* = to hold the value / text of the node
+2. isword: *bool* = to mark whether that node in the trie marks the completion of a word
+3. children: *Dict* = to hold pointers to other children nodes
+
+Then the trie gets built by creating a node for each letter and adding it as a child to the node before it
+
+## Building the trie itself
+
+Start by initializing an empty node
+
+```
+class Trie:
+    def __init__(self):
+        self.node = Node(None)
+```
+
+For the insert operation, fetch the starting node, then for every letter in the word, add it to the children of the letter before it. The final node has its `is_word` attribute marked as **True** because we want to be aware of where the word ends
+
+```   
+def insert(self, word: str) -> None:
+    node = self.node
+    for ltr in word:
+        if ltr not in node.children:
+            node.children[ltr] = Node(ltr)
+        node = node.children[ltr]
+    node.is_word=True
+```
+
+*In the code above, the `node` variable starts by holding a reference to the null node, while the `ltr` iterating variable starts by holding the first letter in `word`. This would ensure that `node` is one level ahead of `ltr`. As they are both moved forward in the iterations, `node` will always remain one level ahead of `ltr`*
+
+For the search operation, fetch the starting node, then for every letter in the word, check if it is present in the `children` attribute of the current node. As long as it is present, repeat for the next letter and next node. If during the search process, we find a letter that is not present, then the word does not exist in the trie. If we successfully get to the end of the iteration, then we have found what we are looking for. It is time to return a value
+
+First take a look at the code (ignore the return value)
```

Comment:
"Ignore the return value" is for the reader, because just below it, I explain that there are two scenarios that will determine what to return. However, I have removed it